sourcegraph/sourcegraph
# Fuzzy finder: refactor and implement Figma design

This is the first step towards implementing the design from https://www.figma.com/file/E0vD9UduSp2liVyDb7dplw/Fuzzy-finder-UI-improvements?node-id=1714%3A3559

While working on the code, I was increasingly unhappy with the convoluted state management inside `useFuzzyState`, which has caused app freezes (infinite loops between cyclic hooks).

The new state management inside `useFuzzyState` is significantly simpler to reason about and it should be much easier to add new functionality to this new codestructure.

The "Actions" tab is no longer enabled with `fuzzyFinderAll` since I suspect it will take a long time until we can make it GA (need at least a few dozen actions to make it GA-ready).

## Test plan

See the CI go green. To test the changes manually

- `sg start`
- Use keyboard shortcuts to open fuzzy finder
- Type in some query and make sure the fuzzy results are as expected
- Reload the page and make sure the query is preserved and the same results are rendered
- Use tab/shift-tab to cycle between fuzzy tabs
- Use mouse to click on different fuzzy tabs
- Use keyboard shortcut to toggle "search everywhere" and "Search within repo" when focusing on the all/files/symbols fuzzy tabs.
- Use mouse to select "search everywhere" and "search within repo"




## App preview:

- [Web](https://sg-web-olafurpg-fuzzy-v3.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
